DietPi-Software | Kavita: An open-source comics/mangas/ebooks with a built-in web reader

drahen opened the pull request:
Kavita is a fast, feature rich, cross platform reading server. Built with a focus for being a full solution for all your reading needs.

- Serve up Manga/Webtoons/Comics (cbr, cbz, zip/rar/rar5, 7zip, raw images) and Books (epub, pdf)
- First class responsive readers that work great on any device (phone, tablet, desktop)
- Dark mode and customizable theming support
- Rich Metadata support with filtering and searching
- Ways to group reading material: Collections, Reading Lists (CBL Import), Want to Read
- Ability to manage users with rich Role-based management for age restrictions, abilities within the app, etc
- Rich web readers supporting webtoon, continuous reading mode (continue without leaving the reader), virtual pages (epub), etc
- Full Localization Support

drahen committed:
Add Kavita

drahen committed:
Remove trailing spaces

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -6427,6 +6435,52 @@ _EOF_
+		then
+			case $G_HW_ARCH in
+				1) local arch='arm';;
+				2) local arch='arm';;
```
```suggestion
				1|2) local arch='arm';;
```

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -6427,6 +6435,52 @@ _EOF_
+			esac
+
+			local fallback_url="https://github.com/Kareadita/Kavita/releases/download/v0.7.8/kavita-linux-$arch.tar.gz"
+			Download_Install "$(curl -sSfL 'https://api.github.com/repos/Kareadita/Kavita/releases/latest' | mawk -F\" "/\"browser_download_url\": .*\/kavita-linux-$arch\.tar\.gz\"/{print \$4}")" /opt
```
On reinstall, `/opt/Kavita` can be removed, as configs are stored in `/mnt/dietpi_userdata/kavita`, right? To have it as failsafe as possible, I'd suggest:
1. `Download_Install` to working dir (just skip the `/opt` argument)
2. Remove `/opt/Kavita` if it exists already.
3. Move the new dir in place.

Ah, and I'd prefer `/opt/kavita` to have everything consistently in lower case. Can be renamed when moving the downloaded+extracted dir in place.

drahen replied:
My apologies, after checking, it seems that the config was stored inside `/opt/Kavita/config` folder and I did not found information on the Kavita website/github on how to change the config location. So I will change on how install, reinstall, and uninstall works:

**On install/reinstall:**
1. Create `/opt/kavita` if it's not exists.
2. Remove everything inside `/opt/kavita` except `config` folder.
3. Use `Download_Install` without `/opt` argument
4. Move the extracted files to `/opt/kavita`. Change default port to 2036 (free).
5. Create user _kavita_.
6. Create data folder (ebooks and comics)
7. Set folder/file permissions for `/opt/kavita`, ebooks and comics.
8. Setup Kavita service.
9. Finish.

**On uninstall:**
1. Remove Kavita service.
2. Remove everything inside `/opt/kavita`
3. Finish.

Please let me know what do you think about this. Thank you.

MichaIng replied:
Did some tests:
- Generally it works to have the data dir elsewhere via `WorkingDirectory=/mnt/dietpi_userdata/kavita`. However, it then does not only look for the `config` dir below this working directory, but also for `wwwroot` and `I18N`. I did not find a way to have only the data/config moved to elsewhere.
- But of course we could create a symlink. Those `kavita.db-shm`/`kavita.db-wal` files are read and written to very regularly. In case of a spinning external drive for `/mnt/dietpi_userdata` this causes constant spinning. We created this `dietpi-arr_to_RAM` script to allow moving those to tmpfs and back automatically on system boot/shutdown.
- As obviously Kavita, other than Servarr apps, is not designed to have config/data outside of the executable dir, I think it is not worth it, so let's drop `/mnt/dietpi_userdata/kavita` for now.

I mostly agree with your install steps. See how we do it in case of phpSysInfo or Ampache:
1. Download to current dir
1. Check if `/opt/kavita/config` exists already
2. If so, remove the one from the current dir (it is included in the archive as well) and move the one from the target dir to the download dir instead.
3. Adjust port to 2036
4. Remove the whole `/opt/kavita`
5. Install from current dir to `/opt/kavita`

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -6427,6 +6435,52 @@ _EOF_
+			Create_User -g dietpi -d /mnt/dietpi_userdata/kavita kavita
+
+			# Data
+			G_EXEC mkdir -p /mnt/dietpi_userdata/{ebooks,comics}
```
```suggestion
			G_EXEC mkdir -p /mnt/dietpi_userdata/{kavita,ebooks,comics}
```
The `Create_User` command does not create the home dir.

drahen replied:
Do you think we should create a home folder for _kavita_ user?

MichaIng replied:
As of my other comment, let's not use a dedicated home/data/config dir for now.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -6427,6 +6435,52 @@ _EOF_
+[Service]
+SyslogIdentifier=Kavita
+User=kavita
+WorkingDirectory=/opt/Kavita
```
```suggestion
WorkingDirectory=/mnt/dietpi_userdata/kavita
```
Not sure whether all configs are written to the executing users home dir, but if the working directory is used at all for any variable files, then those should be `/mnt/dietpi_userdata/kavita`, so that `/opt/kavita` can be safely removed on reinstalls without loosing any configs/data.

drahen replied:
So I guess `/mnt/dietpi_userdata/kavita` will not be used for anything. Can I keep the `WorkingDirectory` as `/opt/kavita`?

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -6427,6 +6435,52 @@ _EOF_
+[Unit]
+Description=Kavita Server (DietPi)
+Wants=network-online.target
+After=network-online.target remote-fs.target
```
```suggestion
After=network-online.target remote-fs.target
StartLimitIntervalSec=60
StartLimitBurst=5
```
I'm not a big fan of `Restart=on-failure`. If a service crashes, then in 95% of the cases a restart will just lead to another crash and repeating it makes things only worse + masks the relevant logs with the repeated restart attempts. Usually if there is a crash, then admins should investigate and fix the underlying reason instead of having the service in an unlimited restart loop. However, if this is seen as necessary here, then at least there must be a limit: max 5 restarts in 60 seconds.

There are some exceptions where services use the systemd restart feature for their internal updaters, i.e. the updater terminates the main process with an error exit code, and systemd then restarts it automatically, while otherwise the service would stay down. If this is the case here, then better would be to trigger the restart on the exact exit code, rather than all types of process failures.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -613,6 +613,14 @@ Available commands:
+		aSOFTWARE_DESC[$software_id]='open source comics/mangas/ebooks media server with a built-in web reader'
+		aSOFTWARE_CATX[$software_id]=2
+		aSOFTWARE_DOCS[$software_id]='https://dietpi.com/docs/software/media/#kavita'
+		# - No RISC-V support yet
```
```suggestion
		# - No RISC-V support yet: https://github.com/Kareadita/Kavita/releases
```
Just to make it easier checking this back in the future 🙂.

drahen committed:
Apply suggestions from code review

Co-authored-by: MichaIng <micha@dietpi.com>

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -6427,6 +6435,53 @@ _EOF_
 
+		if To_Install 212 kavita # Kavita
+		then
+			case $G_HW_ARCH in
```
```suggestion
			# .NET dependency: https://github.com/dotnet/core/blob/main/Documentation/linux-prereqs.md
			case $G_DISTRO in
				5) aDEPS=('libicu63');;
				6) aDEPS=('libicu67');;
				*) aDEPS=('libicu72');;
			esac

			case $G_HW_ARCH in
```

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -6427,6 +6435,53 @@ _EOF_
+			Download_Install "$(curl -sSfL 'https://api.github.com/repos/Kareadita/Kavita/releases/latest' | mawk -F\" "/\"browser_download_url\": .*\/kavita-linux-$arch\.tar\.gz\"/{print \$4}")" /opt
+
+			# User
+			Create_User -g dietpi -d /mnt/dietpi_userdata/kavita kavita
```
As we do not create/use `/mnt/dietpi_userdata/kavita`, make the install dir its home dir:
```suggestion
			Create_User -g dietpi -d /opt/kavita kavita
```

MichaIng committed:
v8.21

- DietPi-Patches | Hotfix possible Orange Pi 5 update issue

MichaIng committed:
v8.22

- DietPi-Software | Kavita: Add .NET dependencies
- DietPi-Software | Kavita: Change default port to 2036 to avoid conflict with Shairport Sync
- DietPi-Software | Kavita: On reinstals, preserve config dir bug replace everything else cleanly
- DietPi-Software | Kavita: Use /opt/kavita (lower-case) as install and config/run user home dir
- DietPi-Software | Kavita: Add to dietpi-services and dietpi-survey_report, as well as README credits and changelog entry

drahen committed:
Merge branch 'MichaIng:master' into master

drahen committed:
Merge branch 'dev' into master

MichaIng committed:
v8.22

- DietPi-Software | Kavita: Disable for ARMv6 as ARM binaries are ARMv7+ only, expected with .NET core
